### PR TITLE
feat: /admin rest api endpoint 

### PR DESCRIPTION
**Pull request by NagyZoltanPeter**
# Description
/admin rest api endpoint implementation and test

# Changes
- [X] Added new /admin directory for endpoint implementation
- [X] Added new test_rest_admin
- [X] common code from json-rpc implementation is extracted and moved to waku_core 

## How to test

1. run tests/wakunode_rest/test_rest_admin.nim

## Issue
https://github.com/waku-org/nwaku/issues/2075

### CODE REVIEW

File: tests/wakunode_rest/test_rest_admin.nim [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    let restPort = Port(58011)
+    let restAddress = ValidIpAddress.init("127.0.0.1")
+    restServer = RestServerRef.init(restAddress, restPort).tryGet()
+
+    installAdminApiHandlers(restServer.router, node1)
+
+    restServer.start()
+
+    client = newRestHttpClient(initTAddress(restAddress, restPort))
```
Comment (SionoiS) [inline tests/wakunode_rest/test_rest_admin.nim:52]: How does that work exactly? Is it call before each tests or once before.

File: waku/waku_core/multiaddrstr.nim
Comment (SionoiS) [inline waku/waku_core/multiaddrstr.nim:1]: If it's only used in `waku_api` why not put it there?

**Review by SionoiS — APPROVED** [review]
LGTM!

Thanks.

File: tests/wakunode_rest/test_rest_admin.nim (hunk 1 above)
Comment (NagyZoltanPeter) [reply tests/wakunode_rest/test_rest_admin.nim:52]: Exactly, it is before/after each test.

File: waku/waku_core/multiaddrstr.nim
Comment (NagyZoltanPeter) [reply waku/waku_core/multiaddrstr.nim:1]: I had the impression it is more an utility for peer management. Not related directly to the api but used only there now.

File: tests/wakunode_rest/test_rest_admin.nim [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  var node1 {.threadvar.}: WakuNode
+  var node2 {.threadvar.}: WakuNode
+  var node3 {.threadvar.}: WakuNode
+  var peerInfo2 {.threadvar.}: RemotePeerInfo
+  var peerInfo3 {.threadvar.}: RemotePeerInfo
+  var restServer {.threadvar.}: RestServerRef
+  var client{.threadvar.}: RestClientRef
+
+  asyncSetup:
+    node1 = newTestWakuNode(generateSecp256k1Key(), ValidIpAddress.init("127.0.0.1"), Port(60600))
```
Comment (vpavlin) [inline tests/wakunode_rest/test_rest_admin.nim:35]: Let's not hardcode ports unless necessary - can you please use `Port(0)` instead for waku nodes?

File: waku/waku_api/rest/admin/handlers.nim [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  ./types
+
+export types
+
+logScope:
+  topics = "waku node rest admin api"
+
+const ROUTE_ADMIN_V1_PEERS* = "/admin/v1/peers"
+
+func decodeRequestBody[T](contentBody: Option[ContentBody]) : Result[T, RestApiResponse] =
```
Comment (vpavlin) [inline waku/waku_api/rest/admin/handlers.nim:33]: Is there a reason why we have this function copied in every handler implementation? Sicne it is using generics, it feels like it should be there only once in the code - obviously, feel free to disregard for this particular PR, but might be worth cleaning up the code if there is no explicit reason to have multiple copies of the func kept around

**Review by vpavlin — APPROVED** [review]
Some comments, but generaly LGTM

File: waku/waku_api/rest/admin/handlers.nim [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  let reqBodyData = contentBody.get().data
+
+  let requestResult = decodeFromJsonBytes(T, reqBodyData)
+  if requestResult.isErr():
+    return err(RestApiResponse.badRequest("Invalid content body, could not decode. " &
+                                          $requestResult.error))
+
+  return ok(requestResult.get())
+
+proc installAdminV1GetPeersHandler(router: var RestRouter, node: WakuNode) =
```
Comment (Ivansete-status) [inline waku/waku_api/rest/admin/handlers.nim:50]: This `proc` is very interesting. I wonder if we better return more information in the response so that it is clear which protocols each peer supports.

Something like:
```
{
  "peers_info": [
    {
      "protocol": "store",
      "peers": ["", "", "etc"]
    },
    {
      "protocol": "admin",
      "peers": ["", "..."]
    },
    {
      "protocol": "lightpush",
      "peers": ["", "..."]
    }
  ]
}
```

File: waku/waku_api/rest/admin/openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+              type: array
+              items:
+                type: string
+      responses:
+        '200':
+          description: Ok
+        '400':
+          description: Cannot connect to one or more peers.
+        '5XX':
+          description: Unexpected error.
```
Comment (Ivansete-status) [inline waku/waku_api/rest/admin/openapi.yaml:51]: I wonder if would it make sense to test this at this point or maybe we already are testing this in another tests.

File: waku/waku_core/multiaddrstr.nim [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  {.push raises: [].}
+
+import
+    libp2p/[peerinfo, switch]
+
+import
+  ./peers
+
+
+
```
Comment (Ivansete-status) [inline waku/waku_core/multiaddrstr.nim:13]: nitpick: kindly don't add too many empty lines :)

File: waku/waku_core/multiaddrstr.nim [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+proc constructMultiaddrStr*(wireaddr: MultiAddress, peerId: PeerId): string =
+  # Constructs a multiaddress with both wire address and p2p identity
+  $wireaddr & "/p2p/" & $peerId
```
Comment (Ivansete-status) [inline waku/waku_core/multiaddrstr.nim:16]: nitpick: I think is better to explicitly add `return` so that the code is clearer to new Nim devs. Applies elsewhere.

File: waku/waku_core/multiaddrstr.nim [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  # Constructs a multiaddress with both location (wire) address and p2p identity
+  if peerInfo.listenAddrs.len == 0:
+    return ""
+  constructMultiaddrStr(peerInfo.listenAddrs[0], peerInfo.peerId)
+
+proc constructMultiaddrStr*(remotePeerInfo: RemotePeerInfo): string =
+  # Constructs a multiaddress with both location (wire) address and p2p identity
+  if remotePeerInfo.addrs.len == 0:
+    return ""
+  constructMultiaddrStr(remotePeerInfo.addrs[0], remotePeerInfo.peerId)
```
Comment (Ivansete-status) [inline waku/waku_core/multiaddrstr.nim:28]: nitpick: kindly add one empty line at the end :)

**Review by Ivansete-status — COMMENTED** [review]
This PR is super beautiful! Thanks for it! I've added a few comments/questions :)

File: waku/waku_api/jsonrpc/admin/handlers.nim [hunk 9]
```diff
@@ -14,30 +14,13 @@ import
...
-  if peerInfo.listenAddrs.len == 0:
-    return ""
-  constructMultiaddrStr(peerInfo.listenAddrs[0], peerInfo.peerId)
-
-proc constructMultiaddrStr*(remotePeerInfo: RemotePeerInfo): string =
-  # Constructs a multiaddress with both location (wire) address and p2p identity
-  if remotePeerInfo.addrs.len == 0:
-    return ""
-  constructMultiaddrStr(remotePeerInfo.addrs[0], remotePeerInfo.peerId)
-
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_api/jsonrpc/admin/handlers.nim:40]: It is moved to waku/waku_core/multiaddrstr.nim due to multiple usage of those functions

File: waku/waku_api/rest/admin/handlers.nim (hunk 3 above)
Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/handlers.nim:33]: I have the same feelings about. I plan some reorg in a separate PR soon regarding rest api types and related functions. I would keep it as is for now till that small enhancement PR.

File: waku/waku_api/rest/admin/handlers.nim (hunk 4 above)
Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/handlers.nim:50]: I would do maybe the other way around, group protocols by peers. 
Shall we discuss in a new issue taken?

File: waku/waku_api/rest/admin/openapi.yaml (hunk 5 above)
Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/openapi.yaml:51]: I would rather add such case, with adding one false peer out of three or sthg.

File: waku/waku_core/multiaddrstr.nim (hunk 6 above)
Comment (NagyZoltanPeter) [reply waku/waku_core/multiaddrstr.nim:13]: Sure!

File: waku/waku_core/multiaddrstr.nim (hunk 7 above)
Comment (NagyZoltanPeter) [reply waku/waku_core/multiaddrstr.nim:16]: Sure!

File: tests/wakunode_rest/test_rest_admin.nim (hunk 2 above)
Comment (vpavlin) [reply tests/wakunode_rest/test_rest_admin.nim:35]: Hmm, no, Port(0) really just means "let the OS decide an available port", so there should be no difference from the perspective of this test - and if it is, there is a bug somewhere:)

Let's keep the ports in, but let's take a look at it

File: waku/waku_api/rest/admin/types.nim [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+proc writeValue*(writer: var JsonWriter[RestJson], value: WakuPeer)
+  {.raises: [IOError].} =
+  writer.beginRecord()
+  writer.writeField("multiaddr", value.multiaddr)
+  writer.writeField("protocol", value.protocol)
+  writer.writeField("connected", value.connected)
+  writer.endRecord()
+
+proc readValue*(reader: var JsonReader[RestJson], value: var WakuPeer)
```
Comment (gabrielmer) [inline waku/waku_api/rest/admin/types.nim:34]: Where are the procs `readValue` and `writeValue` used?

File: waku/waku_api/rest/admin/client.nim [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  return ok(decoded.get())
+
+# proc decodeBytes*(t: typedesc[WakuPeer], data: openArray[byte],
+#                   contentType: Opt[ContentTypeData]): RestResult[WakuPeer] =
+#   if MediaType.init($contentType) != MIMETYPE_JSON:
+#     error "Unsupported response contentType value", contentType = contentType
+#     return err("Unsupported response contentType")
+
+#   let decoded = ?decodeFromJsonBytes(WakuPeer, data)
+#   return ok(decoded)
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/client.nim:45]: Ah, another left over probably, I will check.

File: waku/waku_api/rest/admin/types.nim (hunk 10 above)
Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/types.nim:34]: These are called by JsonReader and JsonWriter classes as "template instantiation".

**Review by jm-clius — APPROVED** [review]
Great to see the last of the "legacy" APIs now in REST.

File: waku/waku_api/rest/admin/handlers.nim (hunk 4 above)
Comment (Ivansete-status) [reply waku/waku_api/rest/admin/handlers.nim:50]: > I would do maybe the other way around, group protocols by peers. Shall we discuss in a new issue taken?

It sounds great to perform "group protocols by peers".
If we don't apply that within this PR, we need to create an issue so that we don't miss that. It's up to you, what you consider better :)
Thanks!

Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/handlers.nim:57]: Ok, its there! @Ivansete-status please check!

File: waku/waku_api/rest/admin/client.nim [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    error "Unsupported response contentType value", contentType = contentType
+    return err("Unsupported response contentType")
+
+  let decoded = decodeFromJsonBytes(seq[WakuPeer], data)
+
+  if decoded.isErr():
+    # return err(fmt("Invalid response from server, could not decode. {decoded.error}"))
+    return err("Invalid response from server, could not decode.")
+
+  return ok(decoded.get())
```
Comment (SionoiS) [inline waku/waku_api/rest/admin/client.nim:36]:
```suggestion
  let decoded = decodeFromJsonBytes(seq[WakuPeer], data).valueOr:
    return err("Invalid response from server, could not decode.")

  return ok(decoded)
```

File: waku/waku_api/rest/admin/handlers.nim [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+proc installAdminV1PostPeersHandler(router: var RestRouter, node: WakuNode) =
+  router.api(MethodPost, ROUTE_ADMIN_V1_PEERS) do (contentBody: Option[ContentBody]) -> RestApiResponse:
+
+    let decodedBody = decodeRequestBody[seq[string]](contentBody)
+
+    if decodedBody.isErr():
+      return RestApiResponse.badRequest(fmt("Failed to decode request: {decodedBody.error}"))
+
+    let peers: seq[string] = decodedBody.value()
```
Comment (SionoiS) [inline waku/waku_api/rest/admin/handlers.nim:104]:
```suggestion
     let peers: seq[string] = decodeRequestBody[seq[string]](contentBody).isOkOr:
      return RestApiResponse.badRequest(fmt("Failed to decode request: {error}"))
```

File: waku/waku_api/rest/admin/handlers.nim [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+    for i, peer in peers:
+      let peerInfo = parsePeerInfo(peer)
+      if peerInfo.isErr():
+        return RestApiResponse.badRequest(fmt("Couldn't parse remote peer info: {peerInfo.error}"))
+
+      let connOk = await node.peerManager.connectRelay(peerInfo.value, source="rest")
+      if not connOk:
+        return RestApiResponse.badRequest(fmt("Failed to connect to peer at index: {i} - {peer}"))
```
Comment (SionoiS) [inline waku/waku_api/rest/admin/handlers.nim:113]:
```suggestion
      let peerInfo = parsePeerInfo(peer).isOkOr:
        return RestApiResponse.badRequest(fmt("Couldn't parse remote peer info: {error}"))

      if not (await node.peerManager.connectRelay(peerInfo, source="rest")):
        return RestApiResponse.badRequest(fmt("Failed to connect to peer at index: {i} - {peer}"))
```

**Review by SionoiS — APPROVED** [review]

File: tests/wakunode_rest/test_rest_admin.nim [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+      getRes.data.len() == 2
+      # Check peer 2
+      getRes.data.anyIt(it.protocols.find(WakuRelayCodec) >= 0 and
+                   it.multiaddr == constructMultiaddrStr(peerInfo2))
+      # Check peer 3
+      getRes.data.anyIt(it.protocols.find(WakuRelayCodec) >= 0 and
+                   it.multiaddr == constructMultiaddrStr(peerInfo3))
+
+  asyncTest "Set wrong peer":
+    # Connect to nodes 2 and 3 using the Admin API
```
Comment (Ivansete-status) [inline tests/wakunode_rest/test_rest_admin.nim:82]: Is this comment for this test?

File: tests/wakunode_rest/test_rest_admin.nim [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+  asyncSetup:
+    node1 = newTestWakuNode(generateSecp256k1Key(), ValidIpAddress.init("127.0.0.1"), Port(60600))
+    node2 = newTestWakuNode(generateSecp256k1Key(), ValidIpAddress.init("127.0.0.1"), Port(60602))
+    peerInfo2 = node2.switch.peerInfo
+    node3 = newTestWakuNode(generateSecp256k1Key(), ValidIpAddress.init("127.0.0.1"), Port(60604))
+    peerInfo3 = node3.switch.peerInfo
+
+    await allFutures(node1.start(), node2.start(), node3.start())
+    await allFutures(node1.mountRelay(), node2.mountRelay(), node3.mountRelay())
```
Comment (Ivansete-status) [inline tests/wakunode_rest/test_rest_admin.nim:42]: I think this is not needed but not a big deal though :)

File: waku/waku_api/jsonrpc/admin/handlers.nim [hunk 17]
```diff
@@ -14,30 +14,13 @@ import
...
-  # Constructs a multiaddress with both location (wire) address and p2p identity
-  if peerInfo.listenAddrs.len == 0:
-    return ""
-  constructMultiaddrStr(peerInfo.listenAddrs[0], peerInfo.peerId)
-
-proc constructMultiaddrStr*(remotePeerInfo: RemotePeerInfo): string =
-  # Constructs a multiaddress with both location (wire) address and p2p identity
-  if remotePeerInfo.addrs.len == 0:
-    return ""
-  constructMultiaddrStr(remotePeerInfo.addrs[0], remotePeerInfo.peerId)
```
Comment (Ivansete-status) [inline waku/waku_api/jsonrpc/admin/handlers.nim:39]: Ideally we should make this in a previous and separate PR. But I don't see a big trouble as it is quite clear the movement

**Review by Ivansete-status — APPROVED** [review]
Wonderful PR! Thanks!
Just added a couple of comments.

File: tests/wakunode_rest/test_rest_admin.nim (hunk 15 above)
Comment (NagyZoltanPeter) [reply tests/wakunode_rest/test_rest_admin.nim:82]: No, of course copy paste error.

File: waku/waku_api/jsonrpc/admin/handlers.nim (hunk 17 above)
Comment (NagyZoltanPeter) [reply waku/waku_api/jsonrpc/admin/handlers.nim:39]: Thank you. I see it is directly related with this /admin endpoint implementation and would like to avoid code dup.

File: tests/wakunode_rest/test_rest_admin.nim (hunk 16 above)
Comment (NagyZoltanPeter) [reply tests/wakunode_rest/test_rest_admin.nim:42]: Why exactly?
Due maybe right, I have taken the base setup from testing json-rpc admin interface and did not questioned that we need relay protocols mounted to have them able to connect to each other for relay.

File: waku/waku_api/rest/admin/handlers.nim (hunk 13 above)
Comment (NagyZoltanPeter) [reply waku/waku_api/rest/admin/handlers.nim:104]: Cool, thanks! I like this more expressive way.

File: tests/wakunode_rest/test_rest_admin.nim (hunk 16 above)
Comment (Ivansete-status) [reply tests/wakunode_rest/test_rest_admin.nim:42]: I thought that was not needed but after checking carefully I saw it should be there :)

Comment (Ivansete-status) [reply tests/wakunode_rest/test_rest_admin.nim:42]: I thought that was not needed but after checking carefully I saw it should be there :)
